#163388138 Add create articles feature

Arusey opened the pull request:
**What does this PR do?**
User should be able to:
-create an article
-get all articles
-get one article
-edit an article they created
-delete an article they created
**How should this be manually tested?**
1. Run the server:
   `python manage.py runserver`
2. From postman:
   
   1. Create an article by sending a `post` request to the '/api/v1/articles/' end point with the data body
   
   ```
   {
       "article": {
       "title": "Test Article",
       "description": "Test description for the article",
       "body": "Test body for the article"
       }
   }
   ```
   
   1. Retrieve all articles by sending a `get` request to the  '/api/v1/articles/' end point
   2. Retrieve a single article by sending a `get` request to the  '/api/v1/articles/`<slug>`/' end point and pass the articles's slug in the url
   3. Update an  article by sending a `put` request to the  '/api/v1/articles/`<slug>`/' end point and pass the articles's slug in the url. Pass the following as the body
   
   ```
   {
   "article": {
       "title": "Updated title",
       "description": "Updated description",
       "body": "Updated body"
   }
   }
   ```
   
   1. Delete an  article by sending a `delete` request to the  '/api/v1/articles/`<slug>`/' end point

**Any background context you want to provide?**
None
**What are the relevant pivotal tracker stories?**
#163388138
**Screenshots**
1. Create Article
![image](https://user-images.githubusercontent.com/28762562/52694401-a444d200-2f7a-11e9-8f5e-1a74aa74c49e.png)
2. Get all articles
![image](https://user-images.githubusercontent.com/28762562/52695703-42866700-2f7e-11e9-9365-63bd71799519.png)

3. Get one article
![image](https://user-images.githubusercontent.com/28762562/52695757-6053cc00-2f7e-11e9-86a0-210cb97e6e8b.png)

4. Update an article
![image](https://user-images.githubusercontent.com/28762562/52696099-38189d00-2f7f-11e9-93ba-cbf64e5793f5.png)
![image](https://user-images.githubusercontent.com/28762562/52696125-4797e600-2f7f-11e9-91d1-9f24c8ce755b.png)

5. Delete an article
![image](https://user-images.githubusercontent.com/28762562/52697105-c130d380-2f81-11e9-8325-ca50f7e0ddb2.png)

Commit:
feat(authors haven): implement create articles
-ensure users can create view, update and delete articles
-ensure endpoints are secured
[Delivers #163388138]

dev-jey commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,63 @@
+        while Article.objects.filter(slug=new_slug).exists():
+            """increase value of slug by one
+            """
+            new_slug = '{}-{}'.format(slug, s)
```
Hi Kevin. Consider using an `f-string` instead of `.format` as much as possible. Check out the formatting here 
https://realpython.com/python-f-strings/

Arusey replied:
done

asirvex commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,63 @@
+from django.db import models
+from django.utils.text import slugify
+
```
Consider eliminating some of these blank lines

Arusey replied:
done

sirghiny commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,57 @@
+    image = CloudinaryField(blank=True, null=True)
+    slug = models.SlugField(db_index=True, max_length=1000, 
+                             unique=True, blank=True)
+    author = models.ForeignKey(User, related_name='article', on_delete=models.CASCADE)
```
Make the `related_name` "articles" because they'll be many.

sirghiny commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,57 @@
+        if not self.slug:
+            self.slug = self.generate_slug()
+        super().save(*args, **kwargs)
+        
```
Eliminate the blank lines.

sirghiny commented on authors/apps/articles/tests/base_test.py:
```diff
@@ -0,0 +1,82 @@
+from django.urls import reverse
+
+
+class BaseTest(APITestCase):
```
You can reuse the `BaseTest` in `authentication` by inheriting it and adding any methods specific to testing articles.

sirghiny commented on authors/apps/articles/tests/test_articles.py:
```diff
@@ -0,0 +1,243 @@
+        message = response.data["detail"]
+        self.assertEqual(message,
+                         "Not found.")
+
```
Also eliminate these blanks.

Arusey replied:
I will do them ASAP

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -0,0 +1,105 @@
+    queryset = Article.objects.all()
+    permission_classes = (IsAuthenticatedOrReadOnly,)
+
+    def post(self, request):
```
This should be in `ArticleDetail`, the view should be for getting a list of articles. You can use regular `APIView` in `ArticleDetail` to have all the functionality.

Arusey replied:
can you please add more color on this one.

sirghiny replied:
The `ArticleList` class should be doing only listing(returning many articles). Article creation is an operation to do with the details of the article and thus should be in the `ArticleDetail` view. You can inherit from `APIView` instead of `RetrieveUpdateDestroyAPIView`.

Arusey replied:
Hi, because in the `ArticleDetail` class I am just getting, updating and/or deleting the article using its unique slug, I therefore, cannot add a post method. alternatively, i have created new class for posting articles `ArticleCreate`. Just need clarification if this is a rational enough implementation of the CRUD

Commit:
feat(authors haven):code refactor for article crud
-use viewsets to implement crud
-refactor tests accordingly
[Delivers #163388138]

sirghiny commented on authors/apps/articles/serializers.py:
```diff
@@ -0,0 +1,37 @@
+        }
+    )
+
+    def retrieve_author(self, obj):
```
If this method isn't utilised, just remove it.

Arusey replied:
oooh i will ASAP

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -0,0 +1,87 @@
+        article = get_object_or_404(queryset, pk=pk)
+        article_data = request.data.get('article', {})
+        serializer = self.serializer_class(
+            data=article_data, partial=True)
```
This should have the `partial=False` and the `partial_update` should have the `partial=True`.

Arusey replied:
Noted

sirghiny commented on authors/apps/articles/views.py:
```diff
@@ -0,0 +1,87 @@
+        article = get_object_or_404(queryset, pk=pk)
+        article.delete()
+        return Response({"message": "article deleted successfully"},
+                        status=status.HTTP_204_NO_CONTENT)
```
Since its a successful `delete`, the status should be `200`, `204` is for a successful `get` that found no data but raised no exception.

Arusey replied:
Noted
